File: leetcode/752-open-the-lock.js
// https://leetcode.com/problems/open-the-lock/

/**
 * @param {string[]} deadends
 * @param {string} target
 * @return {number}
 */
var openLock = function(deadends, target) {

  target = Number(target);
  const source = 0;
  const deadendsVtx = new Set(deadends.map(deadend => Number(deadend)));
  // base cases
  if (deadendsVtx.has(source) || getNeighboors(source, deadendsVtx).length === 0 || getNeighboors(target, deadendsVtx).length === 0) {
    return -1;
  }
  const queue = [source];

  let length = 0;
  const visited = new Set([0]);
  while (queue.length > 0) {
    let layerLength = queue.length;
    while (layerLength > 0) {
      const node = queue.shift();
      if (node === target) return length;
      const neighboors = getNeighboors(node, deadendsVtx);
      for (const neighboor of neighboors) {
        if (!visited.has(neighboor)) {
          visited.add(neighboor);
          queue.push(neighboor);
        }
      }
      layerLength -= 1;
    }
    length += 1;
  }


  return -1;
};


const getNeighboors = (vertex, deadendSet) => {
  const digits = [
    Math.floor(vertex / 1000),
    Math.floor((vertex % 1000) / 100),
    Math.floor((vertex % 100) / 10),
    vertex % 10
  ];
  const neighboors = [];

  
  for (let i = 0; i < 4; i += 1) {

    let up = digits.slice();
    let down = digits.slice();

    if (digits[i] === 0) {
      up[i] = 9;
      down[i] = 1;
    } else if (digits[i] === 9) {
      up[i] = 8;
      down[i] = 0;
    } else {
      up[i] -= 1;
      down[i] += 1;
    }

    const upNeighboor = (1000 * up[0] + 100 * up[1] + 10 * up[2] + up[3]);
    const downNeighboor = (1000 * down[0] + 100 * down[1] + 10 * down[2] + down[3]);
    
    if (!deadendSet.has(upNeighboor)) neighboors.push(upNeighboor);
    if (!deadendSet.has(downNeighboor)) neighboors.push(downNeighboor);
  }


  return neighboors;
}

File: leetcode/752-open-the-lock.py
# https://leetcode.com/problems/open-the-lock/

from collections import deque
from typing import List


class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        target = int(target)
        source = 0
        deadends_vtx = set()
        for deadend in deadends:
            deadends_vtx.add(int(deadend))

        # base cases 
        if source in deadends_vtx or len(self.get_neighbors(source, deadends_vtx)) == 0 or len(self.get_neighbors(target, deadends_vtx)) == 0:
            return -1

        queue = deque([source])

        length = 0
        visited = set([source])

        while len(queue) > 0:
            layer_length = len(queue)
            while layer_length > 0:
                node = queue.popleft()
                if node == target:
                    return length
                neighbors = self.get_neighbors(node, deadends_vtx)
                for neighboor in neighbors:
                    if neighboor not in visited:
                        visited.add(neighboor)
                        queue.append(neighboor)
                
                layer_length -= 1
            
            length += 1
        
        return -1
    
    def get_neighbors(self, vertex, deadend_set):
        digits = [
            int(vertex / 1000),
            int((vertex % 1000) / 100),
            int((vertex % 100) / 10),
            vertex % 10
        ]

        neighbors = []
        for i in range(4):
            up = digits[:]
            down = digits[:]

            if digits[i] == 0:
                up[i] = 9
                down[i] = 1
            elif digits[i] == 9:
                up[i] = 8
                down[i] = 0
            else:
                up[i] -= 1
                down[i] += 1
            
            up_neighbor = (1000 * up[0] + 100 * up[1] + 10 * up[2] + up[3])
            down_neighbor = (1000 * down[0] + 100 * down[1] + 10 * down[2] + down[3])

            if up_neighbor not in deadend_set: neighbors.append(up_neighbor)
            if down_neighbor not in deadend_set: neighbors.append(down_neighbor)
        
        return neighbors
